### ISSUE TITLE
Treat blank strings differently to nulls in surrogate_key()

### ISSUE THREAD

**Issue by rjh336** [dbt-labs/dbt-utils#488]
### Describe the bug
If I had a table like so:
| field1 (STRING) | field2 |
| ----- | ----- |
| NULL | 1 |
| '' | 1 |

running `dbt_utils.surrogate_key(['field1', 'field2'])` would produce the same value for both rows

This is because of the coalesce() that happens on each value during concatenation [see here
](https://github.com/dbt-labs/dbt-utils/blob/51ed999a44fcc7f9f502be11e5f190f5bc84ba4b/macros/sql/surrogate_key.sql#L40)

DISTINCT and GROUP BY operations treat null different from '', so should we do that as well when making a unique key? Moreover, what would be the proper way to implement this? You likely cannot use another string as a universal substitute and adding a randomly-generated int or string will break idempotency for surrogate keys.

### System information
```
  - package: dbt-labs/dbt_utils
    version: 0.7.6
```

**Which database are you using dbt with?**
- [ ] postgres
- [ ] redshift
- [x] bigquery
- [ ] snowflake
- [ ] other (specify: ____________)

**The output of `dbt --version`:**
```
0.20.2
```

**Comment by joellabes** [comment]
I agree @rjh336! I wound up overriding this macro internally in a past life. 

It is a change for the better, but we need to do it carefully because surrogate_key is used to generate important unique keys. 

I think we'd save this for a future dbt-utils 1.0 release (which I hope we can make sometime after the changes in https://github.com/dbt-labs/dbt-utils/discussions/487) with a big warning in the release notes.

It would need a new argument `coalesce_nulls_to` which defaults to a string like `'_null_'`. People who already have models that depend on the legacy buggy behaviour could then change that to be `''` in models where it was needed. 

I'd welcome a PR to change this, with the disclaimer that it won't ship for a couple of months. 

As I typed, I realised that there's another, better option: I'd be happy to merge a change for the _next patch version_ of dbt-utils which adds the `coalesce_nulls_to` argument but defaults to `''` (i.e. maintaining the status quo). Combined with clear documentation in the readme that this _will_ change, it would allow people to start opting into their preferred behaviour now instead of having to do it in a rush when 1.0.0 comes out one day.

**Comment by joellabes** [comment]
Just had a horrible thought: I think that coalescing a too-long string into a too-short datatype can lead to it silently truncating (I know it does in MS SQL Server for example). We'd need to have tests to make sure that didn't happen in this case!

**Comment by joellabes** [comment]
OK so we didn't quite make it to getting this changed in advance of 1.0, but better late than never. 

Given that utils 1.0 is going to set the standard API for the [next couple of years](https://github.com/dbt-labs/dbt-core/blob/main/docs/roadmap/2022-08-back-for-more.md#:~:text=in%20dbt%20Cloud.-,Imagining%20dbt%20Core%20v2,-.%20Last%20December) (i.e. we aim for dbt utils 2.0 alongside dbt Core 2.0, more or less), we should optimise for the many invocations of `surrogate_key()` that are still to be written. 

At the same time, hosing people's existing surrogate keys on ideological grounds is no bueno. 

Here's how I think we square this circle: 
- Add a new argument to surrogate_key called `coalesce_nulls_to`, with _no default value_. I imagine that most people won't use this at all, unless they want a split behaviour from one macro to another. This would take precedence over a globally-set var (below), and also suppresses the warning.
- Add a new variable called `surrogate_key_coalesce_nulls_to`. Inside of the surrogate_key() macro, we check for its presence. If it exists, obviously use that value. If it _doesn't_, locally set it to `'_dbt_utils_surrogate_key_null_'` _and_ log a warning. This way, we increase the odds that people notice the fact that the behaviour has changed even if they don't thoroughly read the release notes, BUT they also have a way to mute the warning once they've made a decision.

Approximate code:
```
{% macro surrogate_key(field_list, coalesce_nulls_to) %}
{% set is_coalesce_override_set = coalesce_nulls_to is not None or var('surrogate_key_coalesce_nulls_to', None) is not None %}
  {% if not is_coalesce_override_set %}  
  {% set coalesce_nulls_to = '_dbt_utils_surrogate_key_null_' %}
  {% do log.warn('As of dbt utils 1.0, null values are no longer converted to an empty string. Set the surrogate_key_coalesce_nulls_to variable in dbt_project.yml or coalesce_nulls_to argument in this model to change this behaviour or silence this warning. The {}.{} model triggered this warning.') %}
  {% endif %}

... and so on
{% endmacro %}

````

I would recommend people set their variable to the same default value as we have (`'_dbt_utils_surrogate_key_null_'`). This is just to mute the warnings, without having to make another variable called `stop_warning_me_about_surrogate_key_changes`, and makes somewhat-magic behaviour (the replacement value) explicit. If it feels confusing/overload-y, I'm open to a specific variable to silence the warnings as well. Implementor's choice 💻 

It'd look like this:
```
name: my_dbt_project
version: 1.0.0

config-version: 2

vars:
  dbt_utils: #only available inside of the dbt_utils package
    surrogate_key_coalesce_nulls_to: '_dbt_utils_surrogate_key_null_'
```

**Comment by joellabes** [comment]
Actually, I had a better idea: let the breaking changes **break**. 

- Remove `surrogate_key()` (or, more helpfully, replace it with a macro that just throws an exception, saying that it's been removed)
- Create a new macro `generate_surrogate_key()` which doesn't have any customisability - all nulls have that string named above. 
- Anyone who needs the legacy behaviour can make a copy of the old macro and put it in their project, without the `dbt_utils.` namespace.

**Comment by joellabes** [comment]
Really good point on the downstream impacts of this @b-per! I hadn't given that enough thought. I think it's still workable though: 

- If you are OK with the change in behaviour, you can do a global find and replace for `dbt_utils.surrogate_key` with `dbt_utils.generate_surrogate_key` and you will get the same outcome. Very similar to when we deprecated `not_null_where`, [and had to do some replacements](https://discourse.getdbt.com/t/migrating-away-from-dbt-utils-not-null-where-and-unique-where-tests/2773). 
- If you are not OK with the change in behaviour, then you need to make your own macro, and you're going to find and replace `dbt_utils.surrogate_key` with `surrogate_key`
- Alternatively, you can make your own macro that just passes it through if you want: 

```
-- my_project/macros/surrogate_key.sql
{% macro surrogate_key(field_list) %}
  {{ dbt_utils.generate_surrogate_key(field_list) %}
{% endmacro %}
```

In all of those, it's pretty trivial to update your code (admittedly, if you use the Cloud IDE then someone on your team will need to do a local clone), and you are explicitly doing something to pick which process you use going forward. With my original approach, we were going to change the behaviour and _hope_ people noticed. Which given the tendency for surrogate keys to be used in things like snapshots is not actually a safe idea. 



> I feel like it could also potentially break quite a few projects that don't define an upper band for `dbt_utils`.

If people haven't pinned to `"<1.0.0"` they are in for a bad time; we are explicitly breaking a _lot_ of stuff that has built up over the years, so that we enter the 1.0 era with a clean slate.

**Comment by b-per** [comment]
I am putting myself in the shoes of a dbt Cloud user with dozens of projects and I feel that it will require a lot of work on their side, even though they are not impacted by `surrogate_key` treating `NULL` and `''` the same. As you mentioned, a global find/replace would require cloning all the repos locally. With Option 1 at least they will just get a warning if they don't do anything.

Would it be enough to deter people from upgrading to the next `dbt`/`dbt_utils` version?

**Comment by joellabes** [comment]
If an org using Cloud doesn't have anyone who can make a local clone, perhaps we can make them a GitHub Action to do the F&R in their repo: https://github.com/marketplace/actions/find-and-replace. (Or perhaps global F&R will be part of the new IDE soon?) Regardless, I don't want to hold back one of our platforms because another one is missing an ergonomics feature. 

> they are not impacted by `surrogate_key` treating `NULL` and `''` the same

If they have nulls and no blanks in a surrogate key, they would be impacted by the first approach. All of those surrogate keys would start to have a different value. At best, it just leads to unnecessary snapshot rows being generated and expiring a row too early. At worst, an incremental model doesn't get old rows updated or deleted, and things start getting double-counted. 

> Would it be enough to deter people from upgrading to the next `dbt`/`dbt_utils` version?

Breaking changes tend to dissuade people from upgrading until they're ready. I'm OK with that! At some point the carrots in utils 1.0 (new test types) and future changes that come in 1.1, 1.2 etc (or new features in other packages which also require utils 1.0) will be compelling enough for them to make the switch. 

This doesn't stop them from upgrading their dbt Core version though; the 0.8.x series of dbt utils supports v1.x of Core and will continue to work.

**Comment by callum-mcdata** [comment]
I'll throw my 🎩 in the ring in defense of the breaking changes. Given that this behavior currently fails silently, I think it is worthwhile to begin handling it as the default behavior. 

I think a combination of option 1 and 2 is probably a good compromise? I'd suggest we keep the `surrogate_key` macro and have it just throw a compilation error with all this information contained. Something like:

`WARNING: the macro surrogate_key has been deprecated in dbt_utils v1.0.0. A new macro called generate_surrogate_key has been added with similar functionality but a small change to treat NULLs different from empty strings. This has the potential to re-calculate your surrogate keys. If you want to retain previous behavior, please use the generate_surrogate_key macro with the variable treat_nulls_as_empty_strings set to true`

This way if the user decides to keep the behavior the same as the old functionality it has to be a conscious choice BUT they don't need to create a new macro. The default value of `treat_nulls_as_empty_strings` would be set to false.

**Comment by joellabes** [comment]
Yes @callum-mcdata i think that works well.

Thanks everyone for weighing in, especially as I don't know that I did a good job of explaining the second option. 

Another attempt: 
- The current surrogate_key macro will be fully deprecated and only throw an error message. 
- The new macro, generate_surrogate_key, will default to treating nulls and blanks differently. 
- Users can opt into the old setting at a project-wide level with a variable. 

By doing this, 
- the default path works correctly for all future users without ant additional config, which is the purpose of the v1.0 migration. 
- We also avoid accidentally hosing people's primary keys that they use in snapshots/incremental models, which a) causes data integrity issues and b) could cause a lot of additional spend on reprocessing data. 

The cost of this approach is that anyone using the Cloud IDE who doesn't have someone to do a local clone and update will be unable to upgrade to utils 1.0 without doing something like that GitHub action I linked above. If someone makes the action, we can link it in the changelog at some point.

**Comment by dave-connors-3** [comment]
So just to be clear here, we expect that after 1.0, anyone using the current `surrogate_key()` macro will experience a breaking change, and be forced to update to `generate_surrogate_key()` macro, and if they also don't want their existing keys to change (assuming they have the issue of single empty strings and nulls), they'll need to add a variable to their project to achieve parity?

**Comment by joellabes** [comment]
That's all correct! 

> they'll need to add a variable to their project to achieve parity

There are two options here: 
```yml
# dbt_project.yml
...
vars:
  dbt_utils:
    # Option 1: configurable, so they could also choose a different override 
    # than the one we're using, but that's SO MANY QUOTES!! 
    surrogate_key_coalesce_nulls_to: "''"     

    # Option 2: not configurable, but very easy to get right
    surrogate_key_treat_nulls_as_empty_strings: True
```

I have a 51:49% preference to the boolean one, but implementor's choice.

**Comment by dbeatty10** [comment]
Resolved by #685
